In [2]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

%matplotlib notebook

### 1. import dataset

In [3]:
from sklearn.datasets import load_boston
from sklearn.preprocessing import normalize # to standardize our data
from sklearn.model_selection import train_test_split
data, targets = load_boston(True)
data = normalize(data)
targets = targets.reshape((targets.shape[0],1)) # reshape targets to follow our variables
X_train, X_test, y_train, y_test = train_test_split(data, targets, 
                                                    test_size = 0.3, random_state = 42)

In [6]:
boston = load_boston()

boston.feature_names
boston.

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [13]:
print (data.shape)
print (targets.shape)

print (X_train.shape[0])
print (X_test.shape[0])

(506, 13)
(506, 1)
354
152


In summary: we have a dataset with 13 features and 506 observations.

### 2. Build Computing graph

In [14]:
num_observe = data.shape[0]
num_feature = data.shape[1]



x = tf.placeholder(dtype = tf.float32, shape=[None, num_feature])
y = tf.placeholder(dtype = tf.float32, shape=[None, 1])

In [15]:
W = tf.Variable(tf.random_normal([num_feature, 1]))
b = tf.Variable(tf.random_normal([1]))



y_pred = tf.matmul(x, W) + b

loss = tf.reduce_mean(tf.square(y-y_pred))

training = tf.train.GradientDescentOptimizer(learning_rate = 0.5).minimize(loss)

In [26]:
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    print (sess.run(W))

[[ 1.2791563 ]
 [-0.18129672]
 [-0.8660364 ]
 [-0.32327622]
 [-1.4995571 ]
 [-0.54688823]
 [ 0.68787086]
 [ 0.2049409 ]
 [-1.4410965 ]
 [-0.14512208]
 [-1.0788212 ]
 [-1.0785836 ]
 [-0.38703287]]


### 3. Train the graph

In [27]:
with tf.Session() as sess:
    
    num_of_iteration = 5000
    sess.run(tf.global_variables_initializer())
    
    for i in range(num_of_iteration):
        sess.run(training, feed_dict = {x: X_train, y: y_train})
        if (i % 250 == 0):
            print ('the value of loss function:', loss.eval(feed_dict={x: X_train, y: y_train}))
            
    predictions = sess.run(y_pred, feed_dict={x: X_test})
    differences = predictions.flatten() - y_test.flatten()
    differences = [abs(x) for x in differences]
    print ("House prices are off by an average of", np.mean(differences), "thousand dollars.")
    
    print (sess.run(W))

the value of loss function: 551.3761
the value of loss function: 64.05807
the value of loss function: 62.01592
the value of loss function: 60.532997
the value of loss function: 59.255264
the value of loss function: 58.08736
the value of loss function: 56.9968
the value of loss function: 55.969044
the value of loss function: 54.99608
the value of loss function: 54.072735
the value of loss function: 53.19525
the value of loss function: 52.360687
the value of loss function: 51.566566
the value of loss function: 50.810658
the value of loss function: 50.09097
the value of loss function: 49.405666
the value of loss function: 48.75299
the value of loss function: 48.13132
the value of loss function: 47.539097
the value of loss function: 46.974857
House prices are off by an average of 4.4104595937226945 thousand dollars.
[[ -45.281307 ]
 [  30.668879 ]
 [ -29.298103 ]
 [   3.6610236]
 [   1.2752678]
 [  33.31273  ]
 [   9.06746  ]
 [ -18.885952 ]
 [   8.991432 ]
 [ -13.625144 ]
 [ -41.520454 ]
